bp-activity-subscription-digest.php

```diff
@@ -200,16 +204,17 @@ function ass_digest_fire( $type ) {
 
 			$activity_message .= ass_digest_format_item_group( $group_id, $activity_ids, $type, $group_name, $group_slug, $user_id );
 			unset( $group_activity_ids[ $group_id ] );
```

Comment:
If we're continuing to unset this, doesn't it mean that `$group_activity_ids` and `$group_activity_ids_original` will always be identical? (This shows how we probably need some automated testing here :-D )

Reply:
They will be identical unless $group_activity_ids changes due to filtering, which is what we're doing. This patch intends to capture and preserve the $group_activity_ids elements which are part of the unfiltered meta value and are not removed by the loop building the digest to the current user.

Reply:
Ah right - sorry for misunderstanding.

This approach feels too specific to the filter. It seems to me that the approach should be: remove those items from the stored array if we have good reason to believe they've been sent. See https://github.com/boonebgorges/buddypress-group-email-subscription/commit/e7031e21cdc707d6bcec95c34dd765dd38ae8f93. The end result is similar but it feels like the logic is more closely tied to the goal of removing sent items.

Can you look this over to see if it accomplishes what you're trying to do?

Reply:
I think you're right, this is a more robust approach and aligning the logic around subtracting a sent items array makes more sense than trying to "repair" a filtered array like I did.

It also occurs to me that a filter could potentially remove some activity IDs from a group but keep the rest of the group's activities. I think both of our solutions would remove the entire group before overwriting user meta which would mean the unfiltered activity would get removed without ever being sent in that case. That may be an exceptionally rare use case but it could happen with e.g. distinct digests for different types of activity in addition to the network-level group filtering.

If you want, I'm open to writing some test cases to explore these possibilities and writing a recursive solution that handles filtering on the activity level although it may take me a while with upcoming travel and all. I see there are already some tests so I'll try to follow existing convention but if you have ideas/suggestions about strategy I am more than happy to hear them.

That said I think your solution works well for our use case, and we're actually already running my patch in production so whatever direction you want to take this is good with me.

Reply:
Thanks @modelm ! 

I'm not sure that any of this is actually testable in its current form. It's likely that it'll have to be abstracted out a bit. Let me take a first crack at it, and once there's one or two functioning tests, I'll hand it over to you for filling out some of the details.

I too will be traveling, so this'll be on the calendar for after the new year. Have a great holiday!